File: Day19/Part2/minimum_window_subsequence.py
# LC 727 — Minimum Window Subsequence
# Two-pointer forward then backward shrinking
from typing import List

def minWindow(s: str, t: str) -> str:
    n, m = len(s), len(t)
    if m == 0:
        return ""
    best = ""
    i = 0
    while True:
        # find first position to try
        pos = s.find(t[0], i)
        if pos == -1:
            break
        # forward scan to match t
        j = pos
        k = 0
        while j < n and k < m:
            if s[j] == t[k]:
                k += 1
            j += 1
        if k < m:
            break  # no further window possible
        end = j - 1
        # backward shrink
        k = m - 1
        j = end
        while j >= pos:
            if s[j] == t[k]:
                k -= 1
                if k < 0:
                    break
            j -= 1
        start = j + 1
        cand = s[start:end+1]
        if not best or len(cand) < len(best):
            best = cand
        i = start + 1  # next attempt moves start forward
    return best

if __name__ == "__main__":
    print(minWindow("abcdebdde", "bde"))  # bcde

File: Day19/Part2/minimum_window_subsequence.cpp
#include <bits/stdc++.h>
using namespace std;

// LC 727 — Minimum Window Subsequence
// Two-pointer: forward to find a window, then backward to shrink

string minWindow(string s, string t){
    int n = (int)s.size(), m = (int)t.size();
    if (m == 0) return "";
    string best = "";
    int i = 0;
    while(true){
        int pos = -1;
        // find first s[pos] == t[0] starting at i
        for(int p=i;p<n;++p){ if (s[p]==t[0]){ pos=p; break; } }
        if (pos == -1) break;
        // forward scan to match t
        int j = pos, k = 0;
        while(j < n && k < m){ if (s[j]==t[k]) ++k; ++j; }
        if (k < m) break; // not found, no more windows
        int end = j - 1;
        // backward shrink
        k = m - 1; j = end;
        while(j >= pos){ if (s[j]==t[k]){ --k; if (k < 0) break; } --j; }
    int start = j;
        string cand = s.substr(start, end - start + 1);
        if (best.empty() || cand.size() < best.size()) best = cand;
        i = start + 1; // move start forward for next search
    }
    return best;
}

int main(){
    cout << minWindow("abcdebdde", "bde") << "\n"; // bcde
    return 0;
}
